nexus/src/cidata.rs

```diff
@@ -0,0 +1,93 @@
+use crate::db::{identity::Resource, model::Instance};
+use fatfs::{FileSystem, FormatVolumeOptions, FsOptions};
+use omicron_common::api::external::Error;
+use serde::Serialize;
+use std::io::{Cursor, Write};
+use uuid::Uuid;
+
+impl Instance {
+    pub fn generate_cidata(&self) -> Result<Vec<u8>, Error> {
+        // cloud-init meta-data is YAML, but YAML is a strict superset of JSON.
+        let meta_data = serde_json::to_vec(&MetaData {
+            instance_id: self.id(),
+            local_hostname: &self.runtime().hostname,
+            public_keys: "", // TODO
+        })
+        .map_err(|_| Error::internal_error("failed to serialize meta-data"))?;
+        let cidata =
+            build_vfat(&meta_data, &self.user_data).map_err(|err| {
+                Error::internal_error(&format!(
+                    "failed to create cidata volume: {}",
+                    err
+                ))
+            })?;
+        Ok(cidata)
+    }
+}
+
+#[derive(Serialize)]
+#[serde(rename_all = "kebab-case")]
+struct MetaData<'a> {
+    instance_id: Uuid,
+    local_hostname: &'a str,
+    public_keys: &'a str,
```

Comment:
shouldn't this be a list of keys or am I confusing this with another field?

Reply:
i should check, it might either be a list, or a newline-delimited string

Reply:
it's both at the same time! https://github.com/canonical/cloud-init/blob/9fabfff5aac6e1b398e18429e6dfad54f59e7f75/cloudinit/sources/__init__.py#L866-L892

i'll update this to be a &[String] i think?

Reply:
:+1: